provider/pkg/resources/resources.go

```diff
@@ -28,15 +26,75 @@ type CloudAPIMetadata struct {
 	Functions map[string]CloudAPIFunction `json:"functions"`
 }
 
-// CloudAPIEndpoint is an endpoint template and the values needed to instantiate the template.
+// CloudAPIEndpoint contains information needed to construct a REST endpoint.
 type CloudAPIEndpoint struct {
+	// SelfLinkProperty is the name of the resource property containing a selfLink.
+	// Example: `selfLink`
+	SelfLinkProperty string `json:"selfLinkProperty,omitempty"`
 	// Template is the template string for a REST endpoint.
-	// Example: v1/projects/{projectsId}/locations/{locationsId}/registries/{registriesId}
+	// Example: https://sqladmin.googleapis.com/v1/projects/{project}/instances/{instance}/sslCerts
 	Template string `json:"template,omitempty"`
 	// Values are the concrete values needed to instantiate the template.
 	Values []CloudAPIResourceParam `json:"values,omitempty"`
 }
 
+// URI constructs a concrete URI value based on the properties of a created resource.
+func (e CloudAPIEndpoint) URI(
+	inputs map[string]interface{},
+	outputs map[string]interface{},
+) (string, error) {
+	if len(e.SelfLinkProperty) > 0 {
+		v, ok := outputs[e.SelfLinkProperty].(string)
+		if !ok {
+			return "", fmt.Errorf("selfLink property %q not found", e.SelfLinkProperty)
+		}
+		return v, nil
+	}
+
+	id := e.Template
+	idParams := e.Values
+	for _, param := range idParams {
+		var propValue string
+		if v, has := evalPropertyValue(inputs, param.SdkName); has {
+			propValue = v
+		} else if v, has := evalPropertyValue(outputs, param.SdkName); has {
+			propValue = v
+		}
+
+		if propValue == "" {
+			return "", fmt.Errorf("property %q/%q not found", param.Name, param.SdkName)
+		}
+
+		// The name property can sometimes contain multiple segments. We only care about the last one
+		// because we flattened the path while building metadata.
+		parts := strings.Split(propValue, "/")
+		propValue = parts[len(parts)-1]
+
+		id = strings.Replace(id, fmt.Sprintf("{%s}", param.Name), propValue, 1)
+	}
+	return id, nil
+}
+
+func evalPropertyValue(values map[string]interface{}, path string) (string, bool) {
```

Comment:
Would be nice to add a few unit tests for this and other aspects called by the provider?

Reply:
Yep, this was a direct copy-paste from a different file, and I didn't include the unit test. I'll see about consolidating the usage.